File: Go/Day-5/problem-1/problem-1-solution.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Rule struct {
	a, b int
}

// Function to check if the update is valid according to the rules
func isUpdateValid(line []int, rules []Rule) bool {
	for _, rule := range rules {
		aPos, bPos := -1, 1000

		// Find positions of both a and b in the line
		for i, page := range line {
			if page == rule.a {
				aPos = i
			}
			if page == rule.b {
				bPos = i
			}
		}

		// If both a and b are present and a comes after b, it's invalid
		if aPos != -1 && bPos != -1 && aPos > bPos {
			return false
		}
	}
	return true
}

func main() {
	// Open the input file
	fileName := "inputdata.txt"
	file, err := os.Open(fileName)
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	var rules []Rule

	// Reading rules until we encounter an empty line (or some separator)
	for {
		var line string
		_, err := fmt.Fscanf(file, "%s\n", &line)
		if err != nil {
			break
		}

		if line == "" { // If we encounter an empty line, stop reading rules
			break
		}

		// Parse the rule in the form of "a|b"
		parts := strings.Split(line, "|")
		if len(parts) == 2 {
			a, _ := strconv.Atoi(parts[0])
			b, _ := strconv.Atoi(parts[1])
			rules = append(rules, Rule{a, b})
		}
	}

	var sum int
	// Now process the updates
	for {
		var line string
		_, err := fmt.Fscanf(file, "%s\n", &line)
		if err != nil {
			break
		}

		// Split the update line into individual page numbers
		parts := strings.Split(line, ",")
		lineNumbers := []int{}
		for _, part := range parts {
			page, _ := strconv.Atoi(part)
			lineNumbers = append(lineNumbers, page)
		}

		// Find the middle element
		midpoint := lineNumbers[len(lineNumbers)/2]

		// Check if the update is valid and add the middle element if valid
		if isUpdateValid(lineNumbers, rules) {
			sum += midpoint
		}
	}

	// Output the sum of middle elements of valid updates
	fmt.Printf("Sum of middle elements of valid updates: %d\n", sum)
}

File: 2024/C/5/problem-1/problem-1-solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define INPUT

#ifdef TEST
#define FILE_NAME "test.txt"
#define NUM_RULES 21
#define NUM_LINES 6
#endif

#ifdef INPUT
#define FILE_NAME "inputdata.txt"
#define NUM_RULES 1176
#define NUM_LINES 202
#endif

typedef struct {
    int a;
    int b;
} Rule;

// Function to check if the update is valid according to the rules
int is_update_valid(int line[], int count, Rule rules[], int rule_count) {
    for (int j = 0; j < rule_count; j++) {
        int a_pos = -1, b_pos = 1000;
        
        for (int k = 0; k < count; k++) {
            if (line[k] == rules[j].a) a_pos = k;
            if (line[k] == rules[j].b) b_pos = k;
        }

        // If both a and b are present and a comes after b, it's invalid
        if (a_pos != -1 && b_pos != -1 && a_pos > b_pos) {
            return 0; // Invalid update
        }
    }
    return 1; // Valid update
}

int main() {
    FILE* f = fopen(FILE_NAME, "r");
    if (!f) {
        perror("Error opening file");
        return 1;
    }

    Rule rules[NUM_RULES];
    for (int i = 0; i < NUM_RULES; i++) {
        int a, b;
        fscanf(f, "%d|%d ", &a, &b);
        rules[i].a = a;
        rules[i].b = b;
    }

    int sum = 0;
    for (int i = 0; i < NUM_LINES; i++) {
        int count = 0;
        int line[100] = {0};

        char delim;
        do {
            fscanf(f, "%d%c", &line[count], &delim);
            count++;
        } while (delim == ',' && !feof(f));

        // Find the middle element
        int midpoint = line[count / 2];

        // Check if the update is valid
        if (is_update_valid(line, count, rules, NUM_RULES)) {
            sum += midpoint; // Add the middle element if valid
        }
    }

    printf("Sum of middle elements of valid updates: %d\n", sum);

    fclose(f);
    return 0;
}
